smorest_crud/access_control/models.py

```diff
@@ -0,0 +1,77 @@
+from typing import Optional, TypeVar, Type, Generic, Union
+
+from flask_sqlalchemy import BaseQuery, Model
+from flask import abort
+
+T = TypeVar("T", bound=Model)
+
+
+class AccessControlQuery(BaseQuery):
+    """Base query class to use for access restriction."""
+
+    def query_for_user(self, user: Type[T]) -> Type["AccessControlQuery"]:
+        """Access control query for the given user instance."""
+        raise NotImplementedError()
+
+
+class AccessControlUser(Generic[T]):
+    """A model mixin to implement access checks for a given model/user.
+
+    Required on all models for views with access checks enabled.
+
+    Example::
+
+        class PetQuery(AccessControlQuery):
+            def query_for_user(self, user) -> "PetQuery":
+                return self.filter_by(owner=user)
+
+        class Pet(Model, AccessControlUser):
+            query_class = PetQuery
+
+            def user_can_read(self, user) -> bool:
+                return self.user_can_write(user) or self.owner.id == user.id
+
+            def user_can_write(self, user) -> bool:
+                return user.is_admin  # only administrators can edit pets
+    """
+    query_class: Type[AccessControlQuery]
+
+    @classmethod
+    def query_for_user(cls, user: "AccessControlUser") -> Optional[AccessControlQuery]:
+        """Filter list of items for `user`, or None if disallowed."""
+        if not hasattr(cls.query, "query_for_user"):
+            return None
+        return cls.query.query_for_user(user)
+
+    @classmethod
+    def get_for_user_or_404(cls: Type[Model], user: Type[T], id_value: Union[str, int], key_attr: str = "extid") -> T:
```

Comment:
`extid` is pretty specific to our uses, wonder if the default should come from config?

Reply:
yeah, good idea. would be better to set `id` for a default value?

Reply:
yeah that'd be reasonable